File: cpp/src/examples/keyczar_tool.py
# Demonstrates KeyczarTool's basic commands and encrypts a short message.
#
# Example: python keyczar_tool.py ~/tmp-dir
#
import os
import sys

import keyczar

PLAINTEXT = 'Secret message'

def RunCommands(keyset_type, keyset_password, rsa_path, rsa_pub_path):
    kt = keyczar.KeyczarTool(keyset_type)
    # Creates a RSA key set for encryption
    kt.CmdCreate(rsa_path, keyczar.KeyPurpose.DECRYPT_AND_ENCRYPT,
                 "MyRSATest", keyczar.KeyczarTool.RSA)
    # Adds a first key
    kt.CmdAddKey(rsa_path, keyczar.KeyStatus.ACTIVE, 0,
                 keyczar.KeyczarTool.PBE, keyset_password)
    # Adds a second key
    version = kt.CmdAddKey(rsa_path, keyczar.KeyStatus.ACTIVE, 0,
                           keyczar.KeyczarTool.PBE, keyset_password)
    # Promotes this last key
    kt.CmdPromote(rsa_path, version)
    # Exports public keys
    kt.CmdPubKey(rsa_path, rsa_pub_path, keyczar.KeyczarTool.PBE,
                 keyset_password)

def EncryptMessage(rsa_pub_path):
    encrypter = keyczar.Encrypter.Read(rsa_pub_path)
    ciphertext = encrypter.Encrypt(PLAINTEXT)
    print 'Plaintext:', PLAINTEXT
    print 'Ciphertext (base64w):', ciphertext
    return ciphertext

def DecryptMessage(keyset_type, keyset_password, rsa_path, ciphertext):
    reader = keyczar.KeysetPBEJSONFileReader(rsa_path, keyset_password)
    crypter = keyczar.Crypter.Read(reader)
    plaintext = crypter.Decrypt(ciphertext)
    return plaintext

if __name__ == '__main__':
    if len(sys.argv) != 2 or not os.path.isdir(sys.argv[1]):
        print >> sys.stderr, "Provide an empty temp directory as argument."
        sys.exit(1)

    rsa_path = os.path.join(sys.argv[1], 'rsa')
    rsa_pub_path = os.path.join(sys.argv[1], 'rsa_pub')
    if os.path.isdir(rsa_path) or os.path.isdir(rsa_pub_path):
        print >> sys.stderr, 'Error:', sys.argv[1], 'is not empty.'
        sys.exit(1)
    os.mkdir(rsa_path)
    os.mkdir(rsa_pub_path)

    keyset_type = keyczar.KeyczarTool.JSON_FILE
    keyset_password = 'cartman'

    RunCommands(keyset_type, keyset_password, rsa_path, rsa_pub_path)

    ciphertext = EncryptMessage(rsa_pub_path)
    plaintext = DecryptMessage(keyset_type, keyset_password, rsa_path,
                               ciphertext)

    assert PLAINTEXT == plaintext, 'mismatch plaintext / decrypted plaintext'

File: cpp/src/examples/keyczar_tool.cc
// Demonstrates use of KeyczarTool commands.
//
// Usage: ./keyczar_tool /path/rsa_empty_path /path/rsa_pub_empty_path
//
#include <iostream>
#include <string>

#include <keyczar/base/scoped_ptr.h>
#include <keyczar/base/stl_util-inl.h>
#include <keyczar/key_purpose.h>
#include <keyczar/key_status.h>
#include <keyczar/keyczar_tool/keyczar_tool.h>

using namespace keyczar;
using namespace keyczar::base;
using namespace keyczar::keyczar_tool;

int RunCmds(const std::string& rsa_path, const std::string& rsa_pub_path) {
  ScopedSafeString password(new std::string("cartman"));

  scoped_ptr<KeyczarTool> kz_tool(new KeyczarTool(KeyczarTool::JSON_FILE));
  if (kz_tool.get() == NULL)
    return 1;

  // Creates a new RSA key set with signing purpose.
  if (!kz_tool->CmdCreate(
          rsa_path,
          KeyPurpose::SIGN_AND_VERIFY,  // key set purpose
          "MyRSASigner",                // key set name
          KeyczarTool::RSA)) {                     // asymmetric cipher type
    std::cerr << "Command 'create' failed." << std::endl;
    return 1;
  }

  const KeyczarTool::KeyEncryption pbe_encryption = KeyczarTool::PBE;

  // Adds a first 'active' key to this key set. This key will be encrypted
  // and the 'encrypted' flag of this key set will be set to true. That means
  // after that operation it will only be possible to add encrypted keys to
  // this key set.
  const int key1_version = kz_tool->CmdAddKey(
      rsa_path,
      KeyStatus::ACTIVE, // key status
      0,                 // use default cipher size
      pbe_encryption,    // key encryption type
      *password);
  if (key1_version <= 0) {
    std::cerr << "Command 'addkey' failed." << std::endl;
    return 1;
  }

  // Adds a second key to this key set.
  if (kz_tool->CmdAddKey(rsa_path, KeyStatus::ACTIVE, 0, pbe_encryption,
                         *password) <= 0) {
    std::cerr << "Command 'addkey' failed." << std::endl;
    return 1;
  }

  // Promotes the first key to primary
  if (!kz_tool->CmdPromote(rsa_path, key1_version)) {
    std::cerr << "Command 'promote' failed." << std::endl;
    return 1;
  }

  // Exports
  if (!kz_tool->CmdPubKey(rsa_path, rsa_pub_path, pbe_encryption, *password)) {
    std::cerr << "Command 'pubkey' failed." << std::endl;
    return 1;
  }

  std::cout << "Successfully executed all commands!" << std::endl;
  return 0;
}

int main(int argc, char** argv) {
  if (argc != 3)
    return 1;

  const std::string rsa_path(argv[1]);
  const std::string rsa_pub_path(argv[2]);
  return RunCmds(rsa_path, rsa_pub_path);
}
